### PR TITLE
many: add DeviceCtx.OperatingMode()

### PR DESCRIPTION
**Pull request by mvo5**
This commit will allow access to the operating mode from the
device context. This will allow us to take certain actions
in snapstate depending on what mode we are in.

The next step will be to avoid trying to install a bootloader
during the seeding in ephemeral modes.

### CODE REVIEW

File: overlord/snapstate/devicectx.go [hunk 1]
```diff
@@ -38,6 +38,9 @@ type DeviceContext interface {
 
 	// ForRemodeling returns whether this context is for use over a remodeling.
 	ForRemodeling() bool
+
+	// OperatingMode return the operating mode (run,install,recover,...)
```
Comment (pedronis) [inline overlord/snapstate/devicectx.go]: missing .

File: overlord/snapstate/snapstatetest/devicectx.go [hunk 2]
```diff
@@ -49,6 +50,10 @@ func (dc *TrivialDeviceContext) ForRemodeling() bool {
 	return dc.Remodeling
 }
 
+func (dc *TrivialDeviceContext) OperatingMode() string {
+	return dc.OpMode
```
Comment (pedronis) [inline overlord/snapstate/snapstatetest/devicectx.go]: we likely want to return "run" if unset no?,  we will never get "" for real

File: overlord/snapstate/snapstatetest/devicectx.go [hunk 3]
```diff
@@ -57,6 +62,11 @@ func MockDeviceModel(model *asserts.Model) (restore func()) {
 	return MockDeviceContext(deviceCtx)
 }
 
+func MockOperatingMode(operatingMode string) (restore func()) {
```
Comment (pedronis) [inline overlord/snapstate/snapstatetest/devicectx.go]: this doesn't seem very useful alone, I would expect one that takes both a model and an operatingMode

File: overlord/devicestate/devicectx.go [hunk 4]
```diff
@@ -34,6 +34,8 @@ func DeviceCtx(st *state.State, task *state.Task, providedDeviceCtx snapstate.De
 	if providedDeviceCtx != nil {
 		return providedDeviceCtx, nil
 	}
+	devMgr := deviceMgr(st)
```
Comment (cmatsuoka) [inline overlord/devicestate/devicectx.go]: nitpick: maybe declare this closer to the place it's actually used?

File: overlord/devicestate/devicectx.go [hunk 5]
```diff
@@ -46,11 +48,16 @@ func DeviceCtx(st *state.State, task *state.Task, providedDeviceCtx snapstate.De
 	if err != nil {
 		return nil, err
 	}
-	return modelDeviceContext{model: modelAs}, nil
+
+	return modelDeviceContext{
+		model:         modelAs,
+		operatingMode: devMgr.OperatingMode(),
```
Comment (cmatsuoka) [inline overlord/devicestate/devicectx.go]: So this will never be empty.

File: overlord/devicestate/devicemgr.go [hunk 6]
```diff
@@ -477,7 +477,7 @@ func (m *DeviceManager) ensureBootOk() error {
 	}
 
 	// book-ok/update-boot-revision is only relevant in run-mode
```
Comment (cmatsuoka) [inline overlord/devicestate/devicemgr.go]: Not part of this commit, but a typo here: book vs boot

File: overlord/devicestate/devicemgr.go [hunk 7]
```diff
@@ -511,7 +511,7 @@ func (m *DeviceManager) ensureInstalled() error {
 	}
 
 	// Note: thisalso stop auto-refreshes indirectly
```
Comment (cmatsuoka) [inline overlord/devicestate/devicemgr.go]: And here, missing a space.

File: overlord/devicestate/devicectx.go [hunk 8]
```diff
@@ -71,3 +78,7 @@ func (dc modelDeviceContext) Store() snapstate.StoreService {
 func (dc modelDeviceContext) ForRemodeling() bool {
```
Comment (pedronis) [inline overlord/devicestate/devicectx.go]: modelDeviceContext has grown enough that we should probably switch to *modelDeviceContext here

File: overlord/devicestate/remodel.go [hunk 9]
```diff
@@ -203,6 +203,7 @@ func remodelCtxFromTask(t *state.Task) (remodelContext, error) {
 
 type baseRemodelContext struct {
 	newModel, oldModel *asserts.Model
+	operatingMode      string
 }
 
 func (rc baseRemodelContext) ForRemodeling() bool {
```
Comment (pedronis) [inline overlord/devicestate/remodel.go]: again, we should switch to *baseRemodelContext if possible

File: overlord/devicestate/devicemgr.go (hunk 7 above)
Comment (pedronis) [reply overlord/devicestate/devicemgr.go]: actually this comment was in the wrong place
